### PR TITLE
Adding support for generating A records for headless services.

### PR DESCRIPTION
**Pull request by vishh**
Unit and e2e tests will be added.

@thockin @ArtfulCoder

### CODE REVIEW

File: cluster/addons/dns/kube2sky/Makefile [hunk 1]
```diff
@@ -4,13 +4,13 @@
 
 .PHONY: all kube2sky container push clean test
 
-TAG = 1.6
+TAG = 1.7
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/Makefile]: We only need to bump it if you intend to cut a release - is that the intent?

**Comment by vishh** [comment]
Sample output:

```shell
# kubectl get services elasticsearch-logging-h
NAME                      LABELS                                                                                                    SELECTOR                     IP(S)     PORT(S)
elasticsearch-logging-h   kubernetes.io/cluster-service=true,kubernetes.io/name=Elasticsearch-headless,name=elasticsearch-logging   name=elasticsearch-logging   None      9200/TCP
```
```shell
$ dig @10.244.1.10 elasticsearch-logging-h.default.cluster.local

;; QUESTION SECTION:
;elasticsearch-logging-h.default.cluster.local. IN A

;; ANSWER SECTION:
elasticsearch-logging-h.default.cluster.local. 30 IN A 10.244.0.4
elasticsearch-logging-h.default.cluster.local. 30 IN A 10.244.1.3
```

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 2]
```diff
@@ -70,43 +83,146 @@ type kube2sky struct {
...
 	}
 	return nil
 }
 
+func (ks *kube2sky) addDNS(subdomain string, service *kapi.Service) error {
+	// if PortalIP is not set, a DNS entry should not be created
+	if !kapi.IsServiceIPSet(service) {
+		return ks.newHeadlessService(subdomain, service)
+	}
+	return ks.generateRecordsForRegularService(subdomain, service)
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: maybe ks.newPortalService() to parallel newHeadlessService() ?

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 3]
```diff
@@ -70,43 +83,146 @@ type kube2sky struct {
...
+			glog.V(2).Infof("Setting DNS record: %v -> %q\n", subdomain, string(b))
+			if err := ks.writeSkyRecord(subdomain, string(b)); err != nil {
+				return err
+			}
 		}
-		b, err := json.Marshal(svc)
+	}
+
+	return nil
+
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: nit: blank line

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 4]
```diff
@@ -70,43 +83,146 @@ type kube2sky struct {
...
+	if err != nil {
+		return err
+	}
+
+	e, exists, err := ks.endpointsStore.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to get endpoints object from endpoints store - %v", err)
+	}
+	if !exists {
+		glog.V(1).Infof("could not find endpoints for service %q in namespace %q", service.Name, service.Namespace)
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: ", will try again soon" ?  But what will trigger a retry?

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 5]
```diff
@@ -204,61 +358,60 @@ func newKubeClient() (*kclient.Client, error) {
...
 	)
-	serviceController.Run(util.NeverStop)
+	go serviceController.Run(util.NeverStop)
+	return serviceStore
+}
+
+func watchEndpoints(kubeClient *kclient.Client, ks *kube2sky) kcache.Store {
+	eStore, eController := kframework.NewInformer(
+		&kcache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: how is this form different from NewListWatchFromClient(...) used for services?  at least a comment, if it is really different

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 6]
```diff
@@ -70,43 +83,146 @@ type kube2sky struct {
...
+	return nil
+
+}
+
+func (ks *kube2sky) addDNSUsingEndpoints(e *kapi.Endpoints) error {
+	ks.mlock.Lock()
+	defer ks.mlock.Unlock()
+	svc, exists := ks.headlessServicesMap[nameNamespace{e.Namespace, e.Name}]
+	if !exists {
+		return nil
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: What does this case mean?  You got an update but don't know about the base object?  Should we handle it?

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 7]
```diff
@@ -70,43 +83,146 @@ type kube2sky struct {
...
+	}
+	// Remove existing DNS entry.
+	subdomain := buildDNSNameString(ks.domain, svc.Namespace, svc.Name)
+	if err := ks.removeDNS(subdomain); err != nil {
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: If we're just going to destroy it and recreate it, why do we need the headlessServicesMap at all?

**Comment by thockin** [comment]
Overall good, a few minor points - could use some helpful comments for poor future me (or future you :)

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 4 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: Ack. An endpoints object creation will trigger a retry.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 6 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: This PR watches all the endpoints. This helps filter out endpoints objects that do not belong to headless services.

Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: If the endpoints object had a object reference to the service it belongs to, this map is probably not necessary.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 7 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: See above comment.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 6 above)
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: It doesn't have an object ref, but it does have teh same name - is that good enough?

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 5 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: Switched to NewListWatchFromClient(...) for endpoints too.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 2 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: Done.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 3 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: Removed

File: cluster/addons/dns/kube2sky/Makefile (hunk 1 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/Makefile]: Not yet. Reverted this change.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 6 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: Thats true. I can use the service store instead of the additional map. I will make that change.

Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: SGTM

**Comment by vishh** [comment]
@thockin: Unit tests and e2e tests added. The e2e test needs to be upgraded to verify the number of A records. But, this PR is safe to land as-is.

File: cluster/addons/dns/kube2sky/kube2sky.go [hunk 8]
```diff
@@ -70,43 +80,163 @@ type kube2sky struct {
...
+	glog.V(2).Infof("Removing %s from DNS", subdomain)
+	resp, err := ks.etcdClient.RawGet(skymsg.Path(subdomain), false, false)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		glog.V(2).Infof("Subdomain %q does not exist in etcd", subdomain)
+		return nil
+	}
+	// TODO: Ignore non-existent subdomains.
```
Comment (thockin) [inline cluster/addons/dns/kube2sky/kube2sky.go]: isn't this TODO done by the above?

File: cmd/kubelet/app/plugins.go [hunk 9]
```diff
@@ -34,6 +34,7 @@ import (
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/iscsi"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/nfs"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/persistent_claim"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/rbd"
```
Comment (thockin) [inline cmd/kubelet/app/plugins.go]: why is the RBD PR merged into this one?

**Comment by thockin** [comment]
LGTM except you have another PR merged in here (ceph RBD)

File: cmd/kubelet/app/plugins.go (hunk 9 above)
Comment (vishh) [inline cmd/kubelet/app/plugins.go]: Rebase issue. Fixed it.

File: cluster/addons/dns/kube2sky/kube2sky.go (hunk 8 above)
Comment (vishh) [inline cluster/addons/dns/kube2sky/kube2sky.go]: Good catch. Had it in there while I was working on the PR. Removed.

**Comment by vishh** [comment]
@thockin: Rebase issue fixed and unwanted TODO removed.

**Comment by thockin** [comment]
LGTM

**Comment by vishh** [comment]
Rebased to head and updated e2e tests to use the e2e framework.

**Comment by lavalamp** [comment]
Maybe I am dense but I can't tell from the PR-- what tag did you build kube2sky with, and where is the configuration change to make the cluster use the new build?

**Comment by vishh** [comment]
I am not making a new kube2sky as part of this release.

**Comment by lavalamp** [comment]
I don't understand-- how can the new e2e test pass then?

**Comment by vishh** [comment]
I ran the e2e manually. If it is imperative to make a new release and use
jenkins e2e to verify this PR, I can make a new release.

**Comment by thockin** [comment]
Jenkins will run the e2e tests - if we don't have a kube2sky container
pushed it can't pass, right?

**Comment by vishh** [comment]
Added a new release of kube2sky.

**Comment by ixdy** [comment]
@thockin Jenkins will run e2e tests post-merge, but it isn't running any tests pre-merge yet - only building them.

**Comment by vishh** [comment]
I ran the e2e again and it passed! @saad-ali

**Comment by thockin** [comment]
Thanks!

**Comment by saad-ali** [comment]
Thanks. Merged. Will keep an eye on Jenkins

**Pull request merged**